In [18]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns 
from  plotly.offline import iplot
import plotly as py 
import plotly.tools as tls 
import cufflinks as cf 
import plotly.graph_objs as go
import plotly.express as px
import warnings
warnings.filterwarnings('ignore')

In [19]:
pwd

'C:\\Users\\HP\\Documents\\DS-ML-IOT\\mini project\\EDA project\\loan_sanction'

In [20]:
loan_data= pd.read_excel('loan_sanction.xlsx')

In [21]:
loan_data

Unnamed: 0   Loan_ID Gender Married Dependents     Education  \
0             0  LP001015   Male     Yes          0      Graduate   
1             1  LP001022   Male     Yes          1      Graduate   
2             2  LP001031   Male     Yes          2      Graduate   
3             3  LP001035   Male     Yes          2      Graduate   
4             4  LP001051   Male      No          0  Not Graduate   
..          ...       ...    ...     ...        ...           ...   
362         362  LP002971   Male     Yes         3+  Not Graduate   
363         363  LP002975   Male     Yes          0      Graduate   
364         364  LP002980   Male      No          0      Graduate   
365         365  LP002986   Male     Yes          0      Graduate   
366         366  LP002989   Male      No          0      Graduate   

    Self_Employed  ApplicantIncome  CoapplicantIncome  LoanAmount  \
0              No             5720                  0         110   
1              No             3076               1500         126   
2              No             5000               1800         208   
3              No             2340               2546         100   
4              No             3276                  0          78   
..            ...              ...                ...         ...   
362           Yes             4009               1777         113   
363            No             4158                709         115   
364            No             3250               1993         126   
365            No             5000               2393         158   
366           Yes             9200                  0          98   

     Loan_Amount_Term  Credit_History Property_Area  
0                 360               1         Urban  
1                 360               1         Urban  
2                 360               1         Urban  
3                 360               1         Urban  
4                 360               1         Urban  
..                ...             ...           ...  
362               360               1         Urban  
363               360               1         Urban  
364               360               1     Semiurban  
365               360               1         Rural  
366               180               1         Rural  

[367 rows x 13 columns]

### Understanding Data

In [22]:
loan_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 367 entries, 0 to 366
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Unnamed: 0         367 non-null    int64 
 1   Loan_ID            367 non-null    object
 2   Gender             367 non-null    object
 3   Married            367 non-null    object
 4   Dependents         357 non-null    object
 5   Education          367 non-null    object
 6   Self_Employed      367 non-null    object
 7   ApplicantIncome    367 non-null    int64 
 8   CoapplicantIncome  367 non-null    int64 
 9   LoanAmount         367 non-null    int64 
 10  Loan_Amount_Term   367 non-null    int64 
 11  Credit_History     367 non-null    int64 
 12  Property_Area      367 non-null    object
dtypes: int64(6), object(7)
memory usage: 37.4+ KB


## Insight

We are going to represent the dataset by loan_data in your project

The data is of loan department having the details of customer like loan_ID, Gender, Married status, Dependents, Education detials graduate or not, employed detials, Income of applicant, Income of coapplicated, what is loan amount, loan amount term, credit history, where property is located urban rural or semiurban.

Total 367 rows and 13 cloumns are present. 

First column is not named. 

In dependents column has null value.

Two dtype are present Integer:- 6 columns and object:- 7 columns

## 1. Check how many duplicate rows are present

In [23]:
loan_data.duplicated().sum()


0

### Insight

There is no duplicate rows are present in loan_data


## 2. If any Duplicates Present, drop those duplicates

There is no dupicate rows present

## 3. Check null values in each column

In [24]:
x=[fea for fea in loan_data.columns if loan_data[fea].isnull().sum()>0]
x

['Dependents']

### Insight

Only "Dependents" columns has null values. 

## 4. Delete the Rows which are having null values.

In [25]:
r=loan_data.dropna(inplace=True)
r

In [26]:
y= [fea for fea in loan_data.columns if loan_data[fea].isnull().sum()>0]
y

[]

In [27]:
loan_data.isnull().sum()

Unnamed: 0           0
Loan_ID              0
Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
dtype: int64

In [31]:
loan_data.shape

(357, 13)

### Insight

There was null value in 'Dependents' column, which we can see is being after using dropna function. And the size to dataset is reduced to 357 from 367

## 5. Impute null values with Mode in the Gender Column

In [28]:
loan_data['Gender'].count()

357

In [13]:
loan_data["Gender"].value_counts()

Male      289
Female     68
Name: Gender, dtype: int64

In [33]:
loan_data["Gender"].isnull().sum()

0

### Insight

No null value is present in the gender column. If we want to replace them we can use the above code,  mode of the gender column is "Male", so we can see that all the values will be replaced by "Male" after using the mode and fillna function

## 6. Impute null values with Mode in Self_Employed Column

In [15]:
loan_data["Self_Employed"].isnull().sum()

0

In [35]:
loan_data['Self_Employed'].value_counts()

No     321
Yes     36
Name: Self_Employed, dtype: int64

### Insight

No null value is present in self_employed column

## 7. Impute null values with Median in LoanAmount Column


In [17]:
loan_data["LoanAmount"].isnull().sum()

0

In [39]:
loan_data['LoanAmount'].count()

357

### Insight

No null value is present in 'LoanAmount' column

## 8. Top 5 Records


In [19]:
loan_data.head(5)

Unnamed: 0   Loan_ID Gender Married Dependents     Education Self_Employed  \
0           0  LP001015   Male     Yes          0      Graduate            No   
1           1  LP001022   Male     Yes          1      Graduate            No   
2           2  LP001031   Male     Yes          2      Graduate            No   
3           3  LP001035   Male     Yes          2      Graduate            No   
4           4  LP001051   Male      No          0  Not Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5720                  0         110               360   
1             3076               1500         126               360   
2             5000               1800         208               360   
3             2340               2546         100               360   
4             3276                  0          78               360   

   Credit_History Property_Area  
0               1         Urban  
1               1         Urban  
2               1         Urban  
3               1         Urban  
4               1         Urban

## 9. Export the data frame in Excel.


In [20]:
loan_data.to_excel('output.xlsx', index=False)

### Insight

Created a data frame in Excel and saved by output name

## 10. Number of Rows and Columns in a DataFrame


In [41]:
loan_data.shape

(357, 13)

In [43]:
num_rows, num_cols = loan_data.shape

print("Number of rows:", num_rows)
print("Number of columns:", num_cols)

Number of rows: 357
Number of columns: 13


## 11. Print all Column names in a DF

In [27]:
loan_data.columns

Index(['Unnamed: 0', 'Loan_ID', 'Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area'],
      dtype='object')

In [44]:
column_names = loan_data.columns

for column in column_names:
    print(column)

Unnamed: 0
Loan_ID
Gender
Married
Dependents
Education
Self_Employed
ApplicantIncome
CoapplicantIncome
LoanAmount
Loan_Amount_Term
Credit_History
Property_Area


## 12. Generate the descriptive statistics of Numerical variables.


In [7]:
loan_data.describe()

Unnamed: 0  ApplicantIncome  CoapplicantIncome  LoanAmount  \
count  367.000000       367.000000         367.000000  367.000000   
mean   183.000000      4805.599455        1569.577657  135.980926   
std    106.088014      4910.685399        2334.232099   60.959739   
min      0.000000         0.000000           0.000000   28.000000   
25%     91.500000      2864.000000           0.000000  101.000000   
50%    183.000000      3786.000000        1025.000000  125.000000   
75%    274.500000      5060.000000        2430.500000  157.500000   
max    366.000000     72529.000000       24000.000000  550.000000   

       Loan_Amount_Term  Credit_History  
count        367.000000           367.0  
mean         342.822888             1.0  
std           64.658402             0.0  
min            6.000000             1.0  
25%          360.000000             1.0  
50%          360.000000             1.0  
75%          360.000000             1.0  
max          480.000000             1.0

### Insight 

Using the describe function we can see the statistic of numerical variables. In this, we get the details of a count, mean, standard derivation, minimum value, 25%, 50%, 75%, and the maximin value.

## 13. Display numeric data type columns in a DF


In [14]:
num_col = [i for i in loan_data.columns if loan_data[i].dtype!='object' ]
num_col
loan_data[num_col]

Unnamed: 0  ApplicantIncome  CoapplicantIncome  LoanAmount  \
0             0             5720                  0         110   
1             1             3076               1500         126   
2             2             5000               1800         208   
3             3             2340               2546         100   
4             4             3276                  0          78   
..          ...              ...                ...         ...   
362         362             4009               1777         113   
363         363             4158                709         115   
364         364             3250               1993         126   
365         365             5000               2393         158   
366         366             9200                  0          98   

     Loan_Amount_Term  Credit_History  
0                 360               1  
1                 360               1  
2                 360               1  
3                 360               1  
4                 360               1  
..                ...             ...  
362               360               1  
363               360               1  
364               360               1  
365               360               1  
366               180               1  

[367 rows x 6 columns]

## 14. Select all data types of columns in a DF except the object data type.


In [11]:
dtype_col = [i for i in loan_data.columns if loan_data[i].dtype!='object' ]
dtype_col

['Unnamed: 0',
 'ApplicantIncome',
 'CoapplicantIncome',
 'LoanAmount',
 'Loan_Amount_Term',
 'Credit_History']

### Insight 

A total of 6 columns are present whose dtype is the object

## 15. Extract all the Records where Self_Employed is equal to Yes.

In [15]:
loan_data[loan_data['Self_Employed'] == 'Yes']

Unnamed: 0   Loan_ID  Gender Married Dependents     Education  \
5             5  LP001054    Male     Yes          0  Not Graduate   
37           37  LP001210    Male     Yes          0      Graduate   
38           38  LP001211    Male      No          0      Graduate   
49           49  LP001270    Male     Yes         3+  Not Graduate   
54           54  LP001312    Male     Yes          0  Not Graduate   
61           61  LP001335    Male     Yes          0      Graduate   
73           73  LP001380    Male     Yes          0      Graduate   
77           77  LP001413    Male      No          0      Graduate   
90           90  LP001475    Male     Yes          0      Graduate   
110         110  LP001584  Female      No          0      Graduate   
132         132  LP001728    Male     Yes          1      Graduate   
143         143  LP001791    Male     Yes          0      Graduate   
144         144  LP001794    Male     Yes          2      Graduate   
147         147  LP001817    Male      No          0  Not Graduate   
155         155  LP001857    Male      No          0  Not Graduate   
156         156  LP001862    Male     Yes          2      Graduate   
195         195  LP002083    Male      No          0      Graduate   
199         199  LP002102    Male     Yes          0      Graduate   
200         200  LP002105    Male     Yes          0      Graduate   
212         212  LP002172    Male     Yes         3+      Graduate   
218         218  LP002195    Male     Yes          1      Graduate   
224         224  LP002256  Female      No          2      Graduate   
231         231  LP002298    Male      No          0      Graduate   
246         246  LP002358    Male     Yes          2      Graduate   
248         248  LP002375    Male     Yes          0  Not Graduate   
271         271  LP002482  Female      No          0      Graduate   
300         300  LP002644    Male     Yes          1      Graduate   
302         302  LP002654  Female      No        NaN      Graduate   
303         303  LP002657    Male     Yes          1  Not Graduate   
306         306  LP002721    Male     Yes          2      Graduate   
328         328  LP002806    Male     Yes         3+      Graduate   
338         338  LP002857    Male     Yes          1      Graduate   
340         340  LP002860    Male     Yes          0      Graduate   
341         341  LP002867    Male      No          0      Graduate   
349         349  LP002891    Male     Yes          0      Graduate   
362         362  LP002971    Male     Yes         3+  Not Graduate   
366         366  LP002989    Male      No          0      Graduate   

    Self_Employed  ApplicantIncome  CoapplicantIncome  LoanAmount  \
5             Yes             2165               3422         152   
37            Yes             2267               2792          90   
38            Yes             5833                  0         116   
49            Yes             8000                250         187   
54            Yes             5293                  0          92   
61            Yes             7016                292         125   
73            Yes             3900               2094         232   
77            Yes             6356                  0          50   
90            Yes             3188               2286         130   
110           Yes             4742                  0         108   
132           Yes             3343               1517         105   
143           Yes            32000                  0         550   
144           Yes            10890                  0         260   
147           Yes             8703                  0         199   
155           Yes             1599               2474         125   
156           Yes             4246               4246         214   
195           Yes             5833                  0         116   
199           Yes             1900               1442          88   
200           Yes             870

In [17]:
loan_data[loan_data['Self_Employed'] == 'Yes'].shape

(37, 13)

### Insight 

After extracting all the record where self_employed is equal to Yes, we get 37 rows. 

## 16. Extract all the Records where Property_area is equal to urban in a DF.


In [16]:
loan_data[loan_data['Property_Area'] == 'Urban']

Unnamed: 0   Loan_ID  Gender Married Dependents     Education  \
0             0  LP001015    Male     Yes          0      Graduate   
1             1  LP001022    Male     Yes          1      Graduate   
2             2  LP001031    Male     Yes          2      Graduate   
3             3  LP001035    Male     Yes          2      Graduate   
4             4  LP001051    Male      No          0  Not Graduate   
..          ...       ...     ...     ...        ...           ...   
356         356  LP002935    Male     Yes          1      Graduate   
357         357  LP002952    Male      No          0      Graduate   
360         360  LP002965  Female     Yes          0      Graduate   
362         362  LP002971    Male     Yes         3+  Not Graduate   
363         363  LP002975    Male     Yes          0      Graduate   

    Self_Employed  ApplicantIncome  CoapplicantIncome  LoanAmount  \
0              No             5720                  0         110   
1              No             3076               1500         126   
2              No             5000               1800         208   
3              No             2340               2546         100   
4              No             3276                  0          78   
..            ...              ...                ...         ...   
356            No             3791               1936          85   
357            No             2500                  0          60   
360            No             8550               4255          96   
362           Yes             4009               1777         113   
363            No             4158                709         115   

     Loan_Amount_Term  Credit_History Property_Area  
0                 360               1         Urban  
1                 360               1         Urban  
2                 360               1         Urban  
3                 360               1         Urban  
4                 360               1         Urban  
..                ...             ...           ...  
356               360               1         Urban  
357               360               1         Urban  
360               360               1         Urban  
362               360               1         Urban  
363               360               1         Urban  

[140 rows x 13 columns]

### Insight 

When we extract where property_area is equal to "Urban" only, we get 140 rows present with this condition.

## 17. Print the number of unique values in the Gender column


In [31]:
y=loan_data['Gender'].nunique()
print("The number of unique values in the Gender column is:", y)

The number of unique values in the Gender column is: 2


### Insight

The unique values in the Gender column are 2 that is "Male" and "Female"